### ISSUE TITLE
Valid BigQuery code is unparsable

### ISSUE THREAD

**Issue by daczarne** [sqlfluff/sqlfluff#5492]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When running SQLFluff on BigQuery SQL a valid expression is returned as `Found unparsable section:`.

### Expected Behaviour

For SQLFluff to recognise this expression as valid BIgQuery code.

```sql
SELECT
  o.global_entity_id
  , o.order_id
  , o.created_date
  , ARRAY((
      SELECT AS STRUCT
        items.id
        , items.quantity
        , items.unit_price
        , items.line_item_total
      FROM o.order.items AS items
    )) AS items
FROM `my_project.my_dataset.orders` AS o
WHERE o.created_date BETWEEN '2023-12-01' AND '2023-12-02'
```

---
Nice to have: for SQLFluff to recognise this expression as invalid (unparsable) BQ code

```sql
SELECT
  o.global_entity_id
  , o.order_id
  , o.created_date
  , ARRAY((
      SELECT AS STRUCT
        items.id
        , items.quantity
        , items.unit_price
        , items.line_item_total
      FROM `o.order.items` AS items
    )) AS items
FROM `my_project.my_dataset.orders` AS o
WHERE o.created_date BETWEEN '2023-12-01' AND '2023-12-02'
```

### Observed Behaviour

I have the following query:

```sql
SELECT
  o.global_entity_id
  , o.order_id
  , o.created_date
  , ARRAY((
      SELECT AS STRUCT
        items.id
        , items.quantity
        , items.unit_price
        , items.line_item_total
      FROM o.order.items AS items
    )) AS items
FROM `my_project.my_dataset.orders` AS o
WHERE o.created_date BETWEEN '2023-12-01' AND '2023-12-02'
```

When running `sqlfluff lint` on this query WITH EVERY RULE EXCLUDED (i.e. just checking that templating and parsing are working correctly on this query), I get the following output

![image](https://github.com/sqlfluff/sqlfluff/assets/16988051/72e46f81-0c15-4f7a-90cd-8582e874e157)

I checked the `.sqlfluff parse` output and found that SQLFluff does not know how to parse the expression `FROM o.order.items AS items`. This is a problem because this is perfectly valid BigQuery code. Moreover, the following options are not valid-BQ code:

```sql
SELECT
  o.global_entity_id
  , o.order_id
  , o.created_date
  , ARRAY((
      SELECT AS STRUCT
        items.id
        , items.quantity
        , items.unit_price
        , items.line_item_total
      FROM order.items AS items
    )) AS items
FROM `my_project.my_dataset.orders` AS o
WHERE o.created_date BETWEEN '2023-12-01' AND '2023-12-02'
```

which also returns the same error. Nor:

```sql
SELECT
  o.global_entity_id
  , o.order_id
  , o.created_date
  , ARRAY((
      SELECT AS STRUCT
        items.id
        , items.quantity
        , items.unit_price
        , items.line_item_total
      FROM `o.order.items` AS items
    )) AS items
FROM `my_project.my_dataset.orders` AS o
WHERE o.created_date BETWEEN '2023-12-01' AND '2023-12-02'
```

which does not return the error, but is not valid BQ code.

---
Seems like the parser is unable to understand that `o` is the alias for the table, `order` is a collection (`STRUCT`) in the table, and `items` a collection-array (`ARRAY<STRUCT< ... >>`) in this table.

### How to reproduce

Run `.sqlfluff lint` on the above query.

> Note: I am running with all rules disabled since I was looking at the parsing stage. But it should be alright to include rules too, it might just add additional output due the default rules.

### Dialect

BigQuery

### Version

- SQL Fluff version: 2.3.5
- Python version: 3.10.1

![image](https://github.com/sqlfluff/sqlfluff/assets/16988051/9e5c404d-2df5-41b7-8ac3-cc52b219cc70)

### Configuration

My config is very long as we list every rule and every configurable parameter in the rules. But the relevant part is as follows:

```cfg
[sqlfluff]
dialect = bigquery
disable_noqa = True
encoding = utf-8
fix_even_unparsable = False
large_file_skip_byte_limit = 0
large_file_skip_char_limit = 0
max_line_length = 200
runaway_limit = 10
sql_file_exts = .sql
templater = jinja
verbose = 2
warn_unused_ignores = False

exclude_rules = AL01, AL02, AL03, AL04, AL05, AL06, AL07, AL08, AM01, AM02, AM03, AM04, AM05, CP01, CP02, CP03, CP04, CP05, AL07, AM02, AM06, AM07, CV01, CV02, CV03, CV04, CV05, CV06, CV07, CV08, CV09, CV10, CV11, JJ01, LT01, LT02, LT03, LT04, LT05, LT06, LT07, LT08, LT09, LT10, LT11, LT12, LT13, RF01, RF02, RF03, RF04, RF05, RF06, ST01, ST02, ST03, ST04, ST05, ST06, ST07, ST08, ST09, TQ01
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by daczarne** [comment]
This issue has been resolved. I don't know exactly in which PR/release, but when updating to SQLFluff version 3.0.6 it is no longer there.
